Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.cpp

```diff
@@ -1086,27 +1068,38 @@ static JSContextRef firstRootFrameJSContext()
     return WKBundleFrameGetJavaScriptContext(firstRootFrame(mainFrame));
 }
 
-void asyncReplyHandler(WKTypeRef, void* context)
+void asyncReplyHandler(WKTypeRef result, void* context)
 {
-    auto function = WTF::adopt(static_cast<Function<void()>::Impl*>(context));
-    function();
+    auto function = WTF::adopt(static_cast<Function<void(WKTypeRef)>::Impl*>(context));
+    function(result);
 }
 
 void postMessageWithAsyncReply(const char* name, JSValueRef callback)
 {
     auto context = firstRootFrameJSContext();
     JSValueProtect(context, callback);
 
-    Function<void()> completionHandler = [callback] () mutable {
+    Function<void(WKTypeRef)> completionHandler = [callback] (WKTypeRef result) mutable {
         auto context = firstRootFrameJSContext();
-        JSObjectCallAsFunction(context, JSValueToObject(context, callback, nullptr), JSContextGetGlobalObject(context), 0, nullptr, nullptr);
+        size_t argumentCount { 0 };
+        JSValueRef* arguments { nullptr };
+        JSValueRef resultJS { nullptr };
+
+        if (result) {
```

Comment:
result can be null when message is `GetAllStorageAccessEntries` and `!page`. In this case it seems we should set argumentCount to 1 and result to an empty or null `JSValue`.
We should probably decide the arguments by message name.

Reply:
I think I disagree.  Calling JSObjectCallAsFunction with 1 argument that is null or 0 arguments is pretty equivalent, and doing so would make the abstraction postMessageWithAsyncReply less useful.  As we continue to expand the use of postMessageWithAsyncReply, though, we may find that you're right and we can do something about it then.

Reply:
Ok if they are equivalent then it's fine. Yeah my idea is generally performing the type check based on what message it is, to make the code more robust.

Reply:
That would make it more robust, but also more verbose.  I think it's fine to have an increasingly generic sendWithAsyncReply wrapper that takes whatever WKTypeRef is in the reply and turn it into a JS object, but I'll keep this in mind, and I'll consider whether the more verbose code would be worth it.  This is just a simple case, so it's hard to tell before more widespread adoption.